network/ata_over_ethernet.ksy

```diff
@@ -0,0 +1,515 @@
+            -orig-id:
+              - Err/Feature
+              - errfeat
+            type: u1
+          - id: sector_count
+            -orig-id:
+              - scnt
+            type: u1
+          - id: command_or_status
+            -orig-id:
+              - Cmd/Status
+              - cmdstat
+            type: u1
+        types:
+          flags:
+            -orig-id: AFlags
+            seq:
+              - id: reserved_80
+                -orig-id: Z
+                type: b1
+                valid: false
+              - id: is_lba48_extended
+                -orig-id:
+                  - E
+                  - AOEAFL_EXT
+                  - Extend
+                type: b1
+                doc: this command is LBA48 extended
+              - id: reserved_40
+                -orig-id: Z
+                type: b1
+                valid: false
+              - id: has_device_register
+                -orig-id:
+                  - D
+                  - AOEAFL_DEV
+                  - Device
+                type: b1
+                doc: as defined in the ATA Device/Head register and is only evaluated when the E bit is set.
+              - id: reserved_20
+                -orig-id: Z
+                type: b1
+                valid: false
+              - id: reserved_10
+                -orig-id: Z
+                type: b1
+                valid: false
+              - id: is_asynchronous
+                -orig-id:
+                  - A
+                  - AOEAFL_ASYNC
+                  - Async
+                type: b1
+                doc: |
+                  Asynchronous requests, that can be queued into cache memory and then responded immediately with the same data as in input. Async write requests are not responded at all. In tha case of power loss or failure the cache can be lost.
+              - id: is_write
+                -orig-id:
+                  - W
+                  - AOEAFL_WRITE
+                  - Write
+                type: b1
+                doc: The data must be written, not read.
+
+  query_config:
+    -orig-id: Conf
+    doc: |
+      In a command message the fields Buffer Count, , and AoE must be set to zero by the client and ignored by the server. The remaining fields may be used to query and set the server’s config string.
+    seq:
+      - id: header
+        type: header
+      - id: config_string
+        #type: str
+        size: header.config_string_length
+        doc: |
+          When querying, may be filled with 0xED, which is not ASCII, so I guess it is just a byte array rather than a string.
+
+    types:
+      header:
+        -orig-id: aoe_cfghdr
+        seq:
+          - id: buffer_size_in_messages
+            -orig-id:
+              - Buffer Count
+              - bufcnt
+            type: u2
+            doc: Count of messages that can be buffered. Messages exceeding buffer are dropped.
+            valid:
+              expr: _ == 0 or 1
```

Comment:
This is guaranteed **not** to work. I don't know if such ambiguous things work in Python (if the syntax `2 < 4 <= 5` works there, then anything could I guess), but not in Kaitai Struct.

What you expect to happen - `(_ == 0) or 1`, or `_ == (0 or 1)`? The `or` operator only works properly on booleans in KS.

---

I just checked that the compiler allows pretty much _any_ (i.e. anyhow invalid) expression in `valid/expr` for some reason. And the problem is not only in the underscore: when I substitute `_` for the actual attribute name (so that there should be no problem in deriving the type), it makes no difference:

```ksy
  - id: a
    type: u1
    valid:
      expr: a == 0 or 2
```

The above compiles to

```javascript
    if (!( ((this.a == 0) || (2)) )) {
      throw new KaitaiStream.ValidationExprError(this.a, this._io, "/seq/0");
    }
```

without a word. AFAIR probably there is some "precompile" class `TypeDetector`/`TypeValidator` in the compiler that handles this and apparently only looks for the expressions in certain YAML keys, so it needs to be adapted for the new key supporting expressions (`valid`).

But actually, the missing type checks of `valid` expressions is the only thing why https://github.com/kaitai-io/kaitai_struct/issues/435 is still open. So it's nothing too surprising that this is silently allowed.

Reply:
`1` was a stub (and I meant it to be considered `true`) to remind me to fill it in. I planned to insert there a condition and have forgotten. In some queries that field (and some other ones, such as firmware version ) must be 0.